source/extensions/filters/http/common/compressor/compressor.cc

```diff
@@ -0,0 +1,354 @@
+  }
+  return Http::FilterTrailersStatus::Continue;
+}
+
+bool CompressorFilter::hasCacheControlNoTransform(Http::HeaderMap& headers) const {
+  const Http::HeaderEntry* cache_control = headers.CacheControl();
+  if (cache_control) {
+    return StringUtil::caseFindToken(cache_control->value().getStringView(), ",",
+                                     Http::Headers::get().CacheControlValues.NoTransform);
+  }
+
+  return false;
+}
+
+bool CompressorFilter::isAcceptEncodingAllowed(const Http::HeaderMap& headers) const {
+  const Http::HeaderEntry* accept_encoding = headers.AcceptEncoding();
+  if (!accept_encoding) {
+    config_->stats().no_accept_header_.inc();
+    return false;
+  }
+
+  using EncPair = std::pair<absl::string_view, float>; // pair of {encoding, q_value}
+  std::vector<EncPair> pairs;
+
+  std::map<std::string, uint32_t> allowed_compressors(config_->registeredCompressors());
+
+  for (const auto token : StringUtil::splitToken(accept_encoding->value().getStringView(), ",",
+                                                 false /* keep_empty */)) {
+    EncPair pair = std::make_pair(StringUtil::trim(StringUtil::cropRight(token, ";")), 1);
+    const auto params = StringUtil::cropLeft(token, ";");
+    if (params != token) {
+      const auto q_value = StringUtil::cropLeft(params, "=");
+      if (q_value != params &&
+          StringUtil::caseCompare("q", StringUtil::trim(StringUtil::cropRight(params, "=")))) {
+        auto result = absl::SimpleAtof(StringUtil::trim(q_value), &pair.second);
+        if (!result) {
+          continue;
+        }
+      }
+    }
+
+    pairs.push_back(pair);
+
+    // Disallow compressors with "q=0".
+    // The reason why we add encodings to "pairs" even with "q=0" is that "pairs" contains
+    // client's expectations and "allowed_compressors" is what the server can handle. Consider
+    // the cases of "Accept-Encoding: gzip;q=0, deflate, *" and "Accept-Encoding: deflate, *"
+    // whereas the server has only "gzip" configured. If we just exclude the encodings with "q=0"
+    // from "pairs" then upon noticing "*" we don't know if "gzip" is acceptable by the client.
+    if (!pair.second) {
+      allowed_compressors.erase(std::string(pair.first));
+    }
+  }
+
+  if (pairs.empty()) {
+    // If the Accept-Encoding field-value is empty, then only the "identity" encoding is acceptable.
+    config_->stats().header_not_valid_.inc();
+    return false;
+  }
+
+  std::sort(pairs.begin(), pairs.end(),
+            [&allowed_compressors](const EncPair& a, const EncPair& b) -> bool {
+              if (a.second == b.second && allowed_compressors.count(std::string(a.first)) &&
+                  allowed_compressors.count(std::string(b.first))) {
+                // In case a user specified more than one encodings with the same quality value
+                // select the one which is registered first in Envoy's config.
+                return allowed_compressors[std::string(a.first)] <
```

Comment:
I would try to think about how to do less work in the sorting function. Actually I'm not sure why you need a total sort. Can't you just do a linear lookup to find the best choice?